README.md
NES graphics editor
===================
by Damian Yerrick

A lot of people making "[retraux]" graphics in the style of 1980s video games want a tool that will help make their graphics conform to the limitations of a particular platform.
How better to enforce NES limits than to make graphics directly on the NES?
This program lets you draw pictures with tiles on an NES or anything that can run an NES emulator.
(Sorry, iDevice owners.)
If you have an NES and a [PowerPak], it also lets you preview your graphics on the real thing and adjust your palette to account for the [artifacts] added by the NES's primitive [NTSC video] encoder.

[retraux]: https://allthetropes.org/wiki/Retraux
[PowerPak]: http://www.retrousb.com/product_info.php?products_id=34
[artifacts]: http://blargg.8bitalley.com/parodius/ntsc-presets/
[NTSC video]: https://wiki.nesdev.com/w/index.php/NTSC_video

Using the editor
----------------
The screen is 30 rows of 32 tiles, each 8 by 8 pixels.
A picture can use up to 256 different tiles.
Each 2 by 2 tile (16 by 16 pixel) area can use one of four color sets in the palette.
All color sets have three colors plus a shared background color.

Make a copy of one of the files in the `sample_savs` folder, and put it where your emulator can find it.
This will depend on where your particular emulator stores saved games.
For a PowerPak, .sav files can be kept anywhere on the CF card.
Then run the editor.
First you'll see a title screen; press Start to get to the tile picker.
If the tile picker is full of garbage, erase the whole picture by holding `B` and Left and pressing `Select`, then waiting ten seconds.

### Tile picker

This screen shows all 256 tiles.
Use the Control Pad to select a tile, press `B` to select a color set, and press `A` to start drawing.
Press `Start` to go to the tile editor, or press `Select` to show a menu that leads to the palette editor and tile optimization tools.
From most other screens, you can press `Select` to return to this screen.

NOTE: If you started with a blank .sav file, or you just erased the picture, you won't see much here until you play around in the tile editor.

### Drawing with tiles

Use the Control Pad to move the cursor.
Press `A` Button to place the selected tile, or hold `A` and move to place
the selected tile multiple times.
Press `B` to pick up the tile and color set under the cursor, and press `B` again to cycle among the four color sets with this tile.
Press `Select` to go back to the tile picker, or hold `Select` and use the Control Pad to quickly select a different tile.

A display in the corner of the screen shows where the cursor is and what tile is selected.
It slides out of the way when you try to draw in the area that it covers.

When you draw, nearby tiles may change colors.
This is because color sets are assigned on a grid twice as large as the tile grid.

Press `Start` to open the map tools menu:

* `Zoom In` opens the tile editor with tiles arranged as they are in the map.
* `Show Color Set Map` shows which color sets are assigned to which areas.
* `Copy Area` lets you copy and paste rectangular blocks of tiles. Move to one corner and press `A`. Move to the other and press `A`. Move the block and press `A` to paste. (Attributes are not copied.) Or press `B` to back up a step, `Start` to return to map tools, or `Select` to return to drawing.
*`Screen Tint` allows changing the "tint" or "emphasis" bits that cause the PPU to apply the equivalent of a [color gel] that darkens certain colors. Some TVs have trouble with tint colors other than `red`, `green`, and `blue`; if your TV loses sync while cycling through the tints, keep pressing `A` until the tint comes back to `none`. Don't try to use tint on an NES with a 2C03 or 2C05 PPU (Famicom Titler, PlayChoice, or early RGB mods) because tint on those systems is so broken that it renders the menus unreadable.

[color gel]: https://en.wikipedia.org/wiki/Color_gel

### Palette editor

Choosing `Edit Color Palette` from the tile picker's menu shows
a form for editing the backdrop color and the three colors that
make up each of the four color sets.
Select a color with the Control Pad, then press `A` to change it.
Left and Right change the hue, while Up and Down change the brightness.
To move to a different color, press `A` again, or release `A` if you
held it while changing the color.
To return to the tile picker, press `Select`.

If you change the background color, the menu color will change
along with it to stay visible.
If you change the colors of a color set used in the picture, everything using that color set changes along with it.
Drawing, changing colors, and drawing more will not allow you to break the NES's color limit.

### Tile editor

The tile editor shows a zoomed view of 30 tiles from the tile sheet or a 6 by 5 tile area of the picture.
The thin lines on the border show the edges of a single tile.
A status bar at the bottom shows where on the 128x128 pixel tile sheet (or the 256x240 pixel picture) the cursor is located, the number of the tile corresponding to this location, which color you are using, how many times this tile is used in the picture, and how many tiles are unused.

If a tile is used multiple times in a picture, your changes will affect all uses of the tile.
If you want to change only one use of a tile in the map, then zoom in on the map and press Start, and any tile you draw on that is used more than once will be copied to a new tile that replaces an unused tile. (The `U` for "unique" next to the count of unused tiles denotes this mode.)

Move the pencil cursor with the Control Pad and press `A` to draw.
Press `B` to pick up the color under the pencil, and `B` to cycle among the four colors.
To scroll the view, move the cursor near the edge.
To return to the tile picker, press `Select`.

### Tile optimization

If you have imported a .sav file from a PPU dump, it will probably have a lot of tiles that the scene doesn't use but other scenes use.
To make room for new tiles that you create, you can remove unused information.
The tile picker's menu gives you several ways to do this:

* `Blank Unused Tiles` replaces any tile that is not used in a screen with a solid pattern.
* `Move Used Tiles to Top` is like <samp>Blank Unused Tiles</samp>, except all the used tiles are moved to the top of the tile sheet.</li>
* `Remove Duplicate Tiles` finds tiles in the tile sheet that are identical, replaces each tile in the screen with the first tile, and then blanks the duplicates.

Importing images
----------------
The editor stores your picture in a battery-backed 64 Kbit SRAM chip, which emulators represent as a 8192 byte .sav file.
A blank .sav file and several sample pictures are included in the `sample_savs` folder.
Once you have a .sav file, put it where your emulator or PowerPak can find it, and then start the program.
The PowerPak is ideal for this because its menu lets you choose one of several .sav files before starting the editor.

Or instead of starting from a blank .sav using the tile editor, you may want to convert existing images.
A conversion tool called written in Python called `savtool.py` is included.
To run it, you'll need to install Python3 and Pillow on your computer.
On Ubuntu or another Debian-based system, installing the `python3-pil` package will get all prerequisites:

    sudo apt-get install python3-pil

If you are running Windows, you'll probably need pre-built executable versions of Python and Pillow. Get Python from python.org; install Pillow through pip3.

For information on the more obscure operations possible with `savtool.py`, run these commands from within a command prompt:

    savtool.py --help
    savtool.py --more-help

### From PPU dump

FCEUX for Windows can export a 16 KiB dump of the PPU memory, which includes two pattern tables (or tile sheets), four screens, and a palette.
To create this dump, from the Debug menu choose Hex Editor, then from the File menu choose Dump to File > PPU Memory.
`savtool.py` can convert this PPU dump to a .sav file for the editor.
Make sure to name the PPU dump with a .ppu suffix.

The PPU's displayable area is two screens tall and two screens wide, though the NES itself has only enough memory for two distinct screens:
two across by one down or one across by two down, depending on how the cartridge board is configured.
Games that scroll over a large map, such as _Contra_ and _Mega Man 2_, continuously load newly visible portions into a ["seam" area] placed just offscreen.
If the scene is not centered, you can change the scrolling.
Scrolling is specified in 16-pixel units, the same size as the area assigned to a color set, and each screen is 16 units wide and 15 units high.

The PPU actually has two tile sheets of 256 tiles each available to it, at addresses $0000 and $1000.
Usually, one is used for the background and the other for sprites.
Most later games use $0000 for backgrounds and $1000 for sprites, but some early games such as _Super Mario Bros._ use the opposite convention.
(The editor itself puts user interface graphics in $0000 and your tile sheet in $1000.)
If the image appears scrambled after conversion, you can use the `--chr` option to tell the converter to pull tiles from $1000 instead.

`savtool.py` does not support raster effects, which are commonly used in NES games to scroll the playfield while holding the status bar still.
For example, a status bar in games like <i>Super Mario Bros.</i> will usually be in the wrong place after conversion.

Example:

    ./savtool.py smb1_title.ppu smb1_title.sav --chr $1000 -x 16 -y 0

["seam" area]: http://wiki.nesdev.com/w/index.php/File:SMB1_scrolling_seam.gif

### From BMP, GIF, or PNG

`savtool` converts a picture or tile sheet in a bitmapped format (.bmp, .gif, or .png) to a .sav file.
If the file is 128x128 pixels or smaller, it loads every tile into the tile sheet and creates a blank nametable.

A larger picture is treated as a full-screen image, and duplicate tiles will be removed.
When converting a bitmap, you can specify an NES palette with `--palette` from either another .sav file or a 32-character hex palette, and `savtool` will do its best to make the colors match given the 16x16 pixel color area restriction.
If you don't provide a `--palette`, the converted .sav uses color set 0 for the whole picture, and you can change the colors of screen areas within the editor by picking up tiles (`B`), changing their color (`B` repeatedly), and putting them back down (`A`).

`savtool` can also render a .sav file to a PNG using Bisqwit's NES palette.
Or if you use `--write-chr` and give a color set number, it extracts the .sav's tile sheet in that color set.
Finally, use `--print-palette` to print the 32-character hex palette stored in a .sav, or `--show` to just view the picture without saving it.

Examples:

    ./savtool.py mysketch.png mysketch.sav
    ./savtool.py mysketch.png mysketch.sav --palette oldsketch.sav
    ./savtool.py mysketch.sav mysketch.png
    ./savtool.py mysketch.sav mysketch.png --write-chr 0
    ./savtool.py mysketch.sav --print-palette
    ./savtool.py mysketch.sav --show

### From CHR

If you have a 4096 byte .chr file and 1024 byte .nam file that you've been using with some other NES graphics editor such as [NESST], you can import those with `savtool`. If no `--palette` is specified, either as a hex code or as another .sav file, it uses the same palette that the editor uses for a blank picture.

    ./savtool.py kitty.chr --palette 0f0010300f0717270f1a2a370f001030 kitty.sav
    ./savtool.py kitty.nam --chr kitty.chr kitty.sav

Converting back to .chr and .nam is also supported:

    ./savtool.py kitty.sav kitty.chr
    ./savtool.py kitty.sav kitty.nam

If you cannot run `savtool.py`, you can convert .chr to .sav by appending 4096 bytes of junk data.

* On Linux and other UNIX-like systems:  
  `cat kitty.chr sample_savs/chr2sav.bin &gt; kitty.sav`
* On Windows:  
  `copy /b kitty.chr+sample_savs/chr2sav.bin kitty.sav`

For more information on the .sav format used by the editor, read "SAV format.txt" in the docs folder.

[NESST]: http://shiru.untergrund.net/software.shtml

Legal
----
The following applies to this manual and the associated programs:

© 2012 Damian Yerrick

Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided
the copyright notice and this notice are preserved in all source
code copies.  This file is offered as-is, without any warranty.


CHANGES.txt
0.06 (2019-11-23)
* Build with recent ca65 (no changes otherwise)
* Map tools: Port to Python 3
* Map tools: Write swatches as .pal, TSV (.txt), or GIMP palette
  (.gpl)
* Map tools: Automatically reduce tile count if greater than 256
  using JRoatch's tile vector quantizer

0.05 (2012-09-29)
* Map editor: Removed duplicate binary-to-decimal converter.
* Map editor: Made some UI code reusable by other map tools.
* Map tools: Added copy and paste (requested by Kasumi).
* Tile editor: Moved everything up 8 pixels to make room for another
  line of text in the status bar.
* Tile editor: Displays count of unused tiles.
* Tile editor: When zoomed in on a map with unused tiles, Start
  toggles allocation of a fresh unused tile when drawing onto a tile
  used more than once.

0.04 (2012-09-13)
* Left+B+Select at the title screen erases the saved data in case
  garbage was present.
* Tile editor: Pen color indicator (at lower right) and tile edge
  marks (on the border) are drawn in the same color as the status bar
  text, fixing contrast with light backgrounds (requested by Kasumi).
* Map editor: Press Start for the map tools menu.
* Map tools: Zoom in on the map and edit tiles in the order they
  appear in the map.
* Map tools: View with constant tiles to see attribute assignments
  (requested by rane).
* Map tools: Change the screen tint (requested by Kasumi).
* savtool: Complete overhaul of image converter.
* savtool: Extract a palette from a .sav as a 32-digit hex code.
* savtool: Convert a color PNG with a specified palette (requested
  by tokumaru).
* savtool: Remap a PNG to use a specific tile sheet.
* savtool: Combine .nam and .chr into .sav or vice versa.
* sample_savs: Cleaned up pm a bit.
* sample_savs: RPG_village has better tree highlights and some
  figures.

0.03 (2012-09-02)
* More completely enables MMC1 SRAM, even on later MMC1 revisions.
* Moved palette editor to top of the tile picker's menu, above the
  help page for the map editor (requested by Kasumi).
* Palette editor supports "spring-loaded" editing.  If the A Button
  is held longer than half a second, or if a color is changed while
  A is held, the editor leaves change mode (requested by Kasumi).
* Palette editor: Fixed wrapping of hues between $xC and $x0
  (requested by Kasumi).
* Map editor: Press B repeatedly to cycle among color sets (requested
  by doppel).
* Map editor: Cursor is a square (requested by rane and Kasumi).
* Tile editor: Drawing a straight line doesn't freeze the top row
  (requested by NovaYoshi).
* Tile editor: First press of B picks up color under pencil point
  (requested by Kasumi).
* Converted manual to HTML and emphasized the pattern of using Select
  as a back button (requested by Kasumi).

0.02 (2012-08-22)
* Added tile editor (press Start from tile picker)
* Includes sample pictures
* Fixed problem with parsing command line in png2sav

0.01 (2012-07-25)
* Initial release


src/paledit.s
;
; Palette editor for NES graphics editor
; Copyright 2012 Damian Yerrick
;
; Copying and distribution of this file, with or without
; modification, are permitted in any medium without royalty provided
; the copyright notice and this notice are preserved in all source
; code copies.  This file is offered as-is, without any warranty.
;
.include "nes.inc"
.include "global.inc"
.importzp das_timer

HOLD_A_THRESHOLD = 30  ; frames

.segment "CODE"

; paledit does not jsr autorepeat, so it can use das_timer the way
; it wants: to count how long A has been held.

.proc paledit
  asl selected_color
  sec
  rol selected_color

  jsr paledit_draw_bg

  ; Set up OAM to show color set 7 in the color set 0 space
  ldx #23
oaminitloop:
  lda oam_for_colorset_7,x
  sta OAM,x
  dex
  bpl oaminitloop

forever:
  ldx #24
  stx oam_used
  jsr paledit_draw_cursor
  ldx oam_used
  jsr ppu_clear_oam

  lda #VBLANK_NMI
  sta PPUCTRL
  lda nmis
:
  cmp nmis
  beq :-

  ; of where is currently pointed
  ldx selected_color
  jsr paledit_draw_hexvalue
  ldx selected_color
  bne no_reload_cs0
  jsr load_colorset_0
no_reload_cs0:

  ldx #0
  ldy #0
  stx OAMADDR
  lda #>OAM
  sta OAM_DMA
  lda #VBLANK_NMI|BG_0000|OBJ_0000
  sec
  jsr ppu_screen_on
  jsr read_pads

  ; Now respond to input
  ; A toggles between selecting a palette entry and changing it
  lda new_keys
  and #KEY_A
  eor select_state
  sta select_state

  ; If A is being held, increase das_timer; otherwise clear it.
  
  bit cur_keys
  bpl clearDASTimer
  inc das_timer
  bpl doneA
  dec das_timer
  jmp doneA
clearDASTimer:
  lda das_timer
  cmp #HOLD_A_THRESHOLD
  bcc notReleaseA
  lda #0
  sta select_state
notReleaseA:
  lda #0
  sta das_timer
doneA:

  lda new_keys
  and #KEY_B
  beq notB
  lda #0
  sta select_state
notB:

  lda select_state
  bne drag_motion
  jsr paledit_cursor_movement
  jmp dpad_done
drag_motion:
  jsr paledit_change_selected_color
  
  ; If a color was changed, and A is being held, set the hold time
  ; to the threshold time so that releasing A leaves change mode.
  bcc dpad_done
  bit cur_keys
  bpl dpad_done
  lda #HOLD_A_THRESHOLD
  sta das_timer
dpad_done:

  lda new_keys
  and #KEY_START|KEY_SELECT
  bne done
  jmp forever
done:
  lsr selected_color
  lsr selected_color
  ldx #SCREEN_CHRPICKER
  rts
.endproc

.proc paledit_cursor_movement
  lda new_keys
  lsr a
  bcc notRight
  lda selected_color
  and #%00000011
  beq notUp
  lda selected_color
  ora #%00000100
  sta selected_color
  rts
notRight:

  lsr a
  bcc notLeft
  lda selected_color
  and #%00011011
  sta selected_color
  rts
notLeft:

  lsr a
  bcc notDown
  lda selected_color
  and #$0B
  cmp #$0B
  beq notUp
  cmp #3
  bne down_not3
  lda selected_color
  clc
  adc #5
  sta selected_color
down_not3:
  inc selected_color
  rts
notDown:

  lsr a
  bcc notUp
  lda selected_color
  and #$0B
  beq notUp
  cmp #1
  bne up_not1
  lda selected_color
  and #$10
  sta selected_color
  rts
up_not1:
  cmp #9
  bne up_not5
  lda selected_color
  sec
  sbc #5
  sta selected_color
up_not5:
  dec selected_color
notUp:
  rts
.endproc

.proc paledit_change_selected_color
  ldx selected_color
  txa
  and #$0F
  beq :+
  tax
:
  lda new_keys
  
  ; Right:
  ; If $xD-$xF, set to $00.
  ; If $xC, set to $x0.
  ; Otherwise, increment.
  lsr a
  bcc notRight
  lda SRAM_PALETTE,x
  and #$0F
  cmp #$0C
  bne not_xC_to_x0
  lda SRAM_PALETTE,x
  and #$30
  sta SRAM_PALETTE,x
  rts
not_xC_to_x0:
  bcs setToDarkGray
  inc SRAM_PALETTE,x
  sec
  rts
setToDarkGray:
  lda #$00
  sta SRAM_PALETTE,x
  rts
notRight:

  ; Left:
  ; If $xD-$xF, set to $00.
  ; If $x0, set to $x0.
  ; Otherwise, decrement.
  lsr a
  bcc notLeft
  lda SRAM_PALETTE,x
  and #$0F
  bne not_x0_to_xC
  lda SRAM_PALETTE,x
  ora #$0C
  sta SRAM_PALETTE,x
  rts
not_x0_to_xC:
  cmp #$0D
  bcs setToDarkGray
  dec SRAM_PALETTE,x
  sec
  rts
notLeft:

  ; Down:
  ; If $0x, set to $0F.
  ; Otherwise, subtract $10.
  lsr a
  bcc notDown
  lda SRAM_PALETTE,x
  sec
  sbc #$10
  bcs :+
  lda #$0F
  sec
:
  sta SRAM_PALETTE,x
  rts
notDown:

  ; Up:
  ; If $xD-$xF, set to $00.
  ; If $3x, set to $30.
  ; Otherwise, add $10.
  lsr a
  bcc notUp
  lda SRAM_PALETTE,x
  and #$0F
  cmp #$0D
  bcs setToDarkGray
up_notBlack:
  lda SRAM_PALETTE,x
  adc #$10
  cmp #$3D
  bcc up_notWhite
  lda #$30
up_notWhite:
  sec
  sta SRAM_PALETTE,x
notUp:
  rts
.endproc

.proc paledit_draw_bg
screensrc = 0
  lda #0
  sta PPUCTRL
  sta PPUMASK
  sta select_state
  tay
  ldx #$20
  jsr ppu_clear_nt
  lda #<paledit_screendata
  sta screensrc
  lda #>paledit_screendata
  sta screensrc+1
segloop:
  ldy #0
  lda (screensrc),y
  beq segdone
  and #$80
  beq not_down
  lda #VRAM_DOWN
not_down:
  sta PPUCTRL
  lda (screensrc),y
  sta PPUADDR
  iny
  lda (screensrc),y
  sta PPUADDR
  iny
byteloop:
  lda (screensrc),y
  beq bytedone
  sta PPUDATA
  iny
  bne byteloop
bytedone:
  tya
  sec  ; increment past the 0
  adc screensrc
  sta screensrc
  bcc :+
  inc screensrc+1
:
  jmp segloop
  
segdone:
  ldx #15
load_all_hexvalues:
  jsr paledit_draw_hexvalue
  dex
  bpl load_all_hexvalues

  ; arrow cursor palette (black, gray, white)
  lda #$3F
  sta PPUADDR
  lda #$11
  sta PPUADDR
  lda #$0F
  sta PPUDATA
  lda #$00
  sta PPUDATA
  lda #$20
  sta PPUDATA
  jmp load_colorset_0
.endproc

.proc paledit_draw_hexvalue
  ; 2108: 0
  ; 2188: 1-3
  ; 2198: 5-7
  ; 2248: 9-11
  ; 2258: 13-15
  txa
  and #$0F
  bne not_bgmirror
  tax
not_bgmirror:
  tay   ; X: palette index; Y: palette index mod 16

  ; Set the palette value
  lda #$3F
  sta PPUADDR
  tya
  beq not_colorset0to7
  cpy #4
  bcs not_colorset0to7
  ora #$1C
not_colorset0to7:
  sta PPUADDR
  lda SRAM_PALETTE,x
  sta PPUDATA

  cpy #8
  lda #$21
  adc #$00
  sta PPUADDR
  lda hexvalue_addrlo,y
  sta PPUADDR
  bne not_x4x8xC
  rts
not_x4x8xC:
  lda SRAM_PALETTE,x
  and #$3F
.if 0
  jsr puthex
  lda #$23
  sta PPUADDR
  sta PPUADDR
  txa
.endif
  jmp puthex
.endproc

.proc paledit_draw_cursor
cursor_y = 0
cursor_x = 3
  lda selected_color
  and #%00000100
  beq not_rightcol
  lda #128
not_rightcol:
  ora #16
  .if 0
    clc
    adc das_timer
  .endif
  sta cursor_x
  lda selected_color
  and #%00001011
  cmp #8
  bcc not_8toF
  sbc #2
not_8toF:
  cmp #0
  beq in_bgspace
  adc #3-1
in_bgspace:
  asl a
  asl a
  asl a
  adc #63
  sta cursor_y
  
  ldx oam_used
  lda cursor_y
  sta OAM+0,x
  lda cursor_x
  sta OAM+3,x
  lda #0
  sta OAM+2,x
  lda select_state
  bne draw_full_arrow
  lda #7  ; outlined right-pointing arrowhead
  sta OAM+1,x
  txa
  clc
  adc #4
  sta oam_used
  rts
draw_full_arrow:
  lda #6  ; outlined right-pointing arrow stem
  sta OAM+1,x
  lda cursor_y
  sta OAM+4,x
  lda cursor_x
  ora #$06
  sta OAM+7,x
  lda #0
  sta OAM+6,x
  lda #7  ; outlined right-pointing arrowhead
  sta OAM+5,x
  txa
  clc
  adc #8
  sta oam_used
  rts
.endproc

.segment "RODATA"
paledit_screendata:
  .byt $20,$67,"EDIT COLOR PALETTE",0
  .byt $20,$E2,"BACKGROUND COLOR",0
  .byt $21,$07,"$",0
  .byt $21,$62,"COLOR SET 0",0
  .byt $A1,$87,"$$$",0
  .byt $21,$72,"COLOR SET 1",0
  .byt $21,$94,1,1," $",0
  .byt $21,$B4,2,2," $",0
  .byt $21,$D4,3,3," $",0
  .byt $22,$22,"COLOR SET 2",0
  .byt $22,$44,1,1," $",0
  .byt $22,$64,2,2," $",0
  .byt $22,$84,3,3," $",0
  .byt $22,$32,"COLOR SET 3",0
  .byt $22,$54,1,1," $",0
  .byt $22,$74,2,2," $",0
  .byt $22,$94,3,3," $",0
  .byt $23,$DD,$11,$00
  .byt $23,$E1,$20,$00
  .byt $23,$E9,$02,$00
  .byt $23,$E5,$30,$00
  .byt $23,$ED,$03,$00
  .byt $23,$03,"SELECT: EXIT",0
  .byt 0
oam_for_colorset_7:
  .byt  95,1,3,32, 95,1,3,40
  .byt 103,2,3,32,103,2,3,40
  .byt 111,3,3,32,111,3,3,40
hexvalue_addrlo:
  .byt $08,$88,$A8,$C8
  .byt $00,$98,$B8,$D8
  .byt $00,$48,$68,$88
  .byt $00,$58,$78,$98

